File: Python/chapter_12.py
# -*- coding: utf-8 -*-
# ---
# jupyter:
#   jupytext:
#     cell_metadata_filter: -all
#     formats: ipynb,py:percent
#     text_representation:
#       extension: .py
#       format_name: percent
#       format_version: '1.3'
#       jupytext_version: 1.19.1
# ---

# %%
"""
This script is a Python conversion of the R script chapter_12.R.
It focuses on t-tests, their relationship with permutation tests, and statistical power.
"""

# %%
import pandas as pd
import numpy as np
from plotnine import ggplot, aes, geom_boxplot, geom_histogram, geom_vline, facet_grid, geom_density
from scipy.stats import ttest_ind, ttest_1samp, t
from statsmodels.stats.power import ttest_power

# %%
np.random.seed(1234)

# %%
# Load and prepare data
try:
    uefa_big5_match_results = pd.read_csv('R/19_20_uefa_big_5.csv')
except FileNotFoundError:
    print("Could not find 'R/19_20_uefa_big_5.csv'.")
    uefa_big5_match_results = pd.DataFrame()

# %%
if not uefa_big5_match_results.empty:
    uefa_big5_match_results['장소'] = pd.Categorical(uefa_big5_match_results['장소'], categories=['안방', '방문'], ordered=True)
    uefa_big5_match_results['시기'] = pd.Categorical(uefa_big5_match_results['시기'], categories=['BC', 'AC'], ordered=True)

    uefa_big5_match_results_period = uefa_big5_match_results[
        (uefa_big5_match_results['장소'] == '안방') &
        (uefa_big5_match_results['리그'] != '리그1')
    ]
    uefa_big5_results_period = uefa_big5_match_results_period.groupby(['팀', '시기'])['승리'].mean().reset_index()
    uefa_big5_results_period = uefa_big5_results_period.rename(columns={'승리': '승률'})

    # Independent samples t-test
    ac_rates = uefa_big5_results_period[uefa_big5_results_period['시기'] == 'AC']['승률']
    bc_rates = uefa_big5_results_period[uefa_big5_results_period['시기'] == 'BC']['승률']
    
    # Welch's t-test (var.equal = False)
    t_stat, p_val = ttest_ind(ac_rates, bc_rates, equal_var=False, alternative='less')
    print(f"Welch's t-test: statistic={t_stat:.3f}, p-value={p_val:.3f}")

    # Student's t-test (var.equal = True)
    t_stat_student, p_val_student = ttest_ind(ac_rates, bc_rates, equal_var=True, alternative='less')
    print(f"Student's t-test: statistic={t_stat_student:.3f}, p-value={p_val_student:.3f}")

    # Paired t-test
    paired_data = uefa_big5_results_period.pivot(index='팀', columns='시기', values='승률').reset_index()
    paired_data = paired_data.dropna(subset=['AC', 'BC'])
    paired_data['diff'] = paired_data['AC'] - paired_data['BC']
    
    t_stat_paired, p_val_paired = ttest_1samp(paired_data['diff'], popmean=0, alternative='less')
    print(f"
Paired t-test: statistic={t_stat_paired:.3f}, p-value={p_val_paired:.3f}")

    # Power analysis
    # First, find the parameters needed
    delta = paired_data['diff'].mean()
    sd_diff = paired_data['diff'].std()
    effect_size = delta / sd_diff # Cohen's d
    
    power = 0.8
    alpha = 0.05
    
    # Calculate required sample size
    required_n = ttest_power(effect_size=effect_size, nobs=None, alpha=alpha, power=power, alternative='smaller')
    print(f"
Required sample size for power=0.8: {np.ceil(required_n)}")
    
    # Using parameters from the R script for NBA data
    # power.t.test(delta = .106, sd = .158, ...)
    nba_delta = 0.106
    nba_sd = 0.158
    nba_effect_size = nba_delta / nba_sd
    required_n_nba = ttest_power(effect_size=nba_effect_size, nobs=None, alpha=alpha, power=power, alternative='one-sided')
    print(f"Required sample size for NBA data: {np.ceil(required_n_nba)}")
    
    # Simulating H0 and H1 distributions for NBA data
    try:
        nba_match_results = pd.read_csv('R/19_20_nba.csv')
    except FileNotFoundError:
        print("
Could not find 'R/19_20_nba.csv'. Skipping H0/H1 simulation.")
        nba_match_results = pd.DataFrame()

    if not nba_match_results.empty:
        nba_bc = nba_match_results[nba_match_results['시기'] == 'BC'].copy()
        nba_bc['장소'] = pd.Categorical(nba_bc['장소'], categories=['안방', '방문'], ordered=True)
        
        nba_bc_summary = nba_bc.groupby(['팀', '장소'])['승리'].mean().reset_index()
        nba_bc_summary = nba_bc_summary.rename(columns={'승리':'승률'})

        # H0 simulation (permutation)
        h0_diffs = []
        for _ in range(1000):
            permuted_labels = np.random.permutation(nba_bc_summary['장소'])
            permuted_df = nba_bc_summary.copy()
            permuted_df['장소'] = permuted_labels
            permuted_means = permuted_df.groupby('장소')['승률'].mean()
            h0_diffs.append(permuted_means['안방'] - permuted_means['방문'])

        nba_simulation_h0 = pd.DataFrame({'stat': h0_diffs, 'type': 'h0'})

        # H1 simulation (bootstrap)
        nba_bc_wide = nba_bc_summary.pivot(index='팀', columns='장소', values='승률').reset_index().dropna()
        nba_bc_wide['차이'] = nba_bc_wide['안방'] - nba_bc_wide['방문']
        
        h1_means = []
        for _ in range(1000):
            bootstrap_sample = nba_bc_wide['차이'].sample(frac=1, replace=True)
            h1_means.append(bootstrap_sample.mean())
        
        nba_simulation_h1 = pd.DataFrame({'stat': h1_means, 'type': 'h1'})

        # Combine and plot
        nba_simulations = pd.concat([nba_simulation_h0, nba_simulation_h1], ignore_index=True)
        
        # Critical value for one-sided test at alpha=0.05
        critical_value = nba_simulation_h0['stat'].quantile(0.95)
        
        power_simulated = (nba_simulation_h1['stat'] > critical_value).mean()
        print(f"
Simulated Power: {power_simulated:.3f}")
        
        p = (ggplot(nba_simulations, aes(x='stat', fill='type')) +
             geom_density(alpha=0.5) +
             geom_vline(xintercept=critical_value, linetype='dashed', size=1))
        # print(p)

# %%
print("
Conversion of chapter_12.R to Python is complete.")

File: notebooks/R/chapter_12.R
# -*- coding: utf-8 -*-
# ---
# jupyter:
#   jupytext:
#     cell_metadata_filter: -all
#     formats: ipynb,R:percent
#     text_representation:
#       extension: .R
#       format_name: percent
#       format_version: '1.3'
#       jupytext_version: 1.19.1
#   kernelspec:
#     display_name: R
#     language: R
#     name: ir
# ---

# %%
pacman::p_load(tidyverse, tidymodels)
set.seed(1234)

# %%
dt(x = 0, df = 1)

# %%
'19_20_uefa_big_5.csv' %>% 
  read.csv() %>% 
  as_tibble -> uefa_big5_match_results

# %%
uefa_big5_match_results %>%
  mutate(
    장소 = 장소 %>% fct_relevel('안방', '방문'),
    시기 = 시기 %>% fct_relevel('BC', 'AC')
  ) -> uefa_big5_match_results

# %%
uefa_big5_match_results %>% 
  filter(장소 == '안방' & 리그 != '리그1') -> uefa_big5_match_results_period  

# %%
uefa_big5_match_results_period %>% 
  group_by(팀, 시기) %>% 
  summarise(승률 = mean(승리), .groups = 'drop') -> uefa_big5_results_period

# %%
uefa_big5_results_period %>% 
  group_by(시기) %>% 
  summarise(승률 = mean(승률), .groups = 'drop') 

# %%
uefa_big5_results_period %>% 
  t_test(formula =  승률  ~  시기,
         order = c('AC', 'BC'),
         alternative = 'less')

# %%
uefa_big5_results_period %>% 
  t_test(formula =  승률  ~  시기,
         order = c('AC', 'BC'),
         alternative = 'less',
         var.equal = TRUE)

# %%
uefa_big5_results_period  %>% 
  group_by(시기) %>% 
  tally()

# %%
uefa_big5_results_period %>% 
  group_by(시기) %>% 
  summarise(평균승률 = mean(승률), 
            분산 = var(승률),
            개수 = n(),
            .groups = 'drop')

# %%
uefa_big5_results_period %>% 
  t_test(formula =  승률  ~  시기,
         order = c('AC', 'BC'),
         alternative = 'less')

# %%
(.421 - .439) / sqrt(.07697/78 + .0337/78)

# %%
(.4213370 - .4388944) / sqrt(.07693693/78 + .03368264/78)

# %%
uefa_big5_results_period %>% 
 specify(승률 ~ 시기) %>%
 hypothesize(null = 'independence') %>%
 generate(reps = 1000, type = 'permute') %>%
 calculate(stat = 'diff in means',
           order = c('AC', 'BC'))

# %%
uefa_big5_results_period %>% 
  specify(승률 ~ 시기) %>%
  hypothesize(null = 'independence') %>%
  calculate(stat = 't',
            order = c('AC', 'BC')) -> uefa_big5_results_period_null_theoretical

# %%
uefa_big5_results_period_null_theoretical

# %%
uefa_big5_results_period_null_theoretical %>% 
  visualize(method = 'theoretical')

# %%
uefa_big5_results_period_null_theoretical %>% 
  visualize(method = 'theoretical') +
  shade_p_value(obs_stat = -.466, direction = 'less')

# %%
uefa_big5_results_period %>% 
  specify(승률 ~ 시기) %>%
  calculate(stat = 't',
            order = c('AC', 'BC')) 

# %%
uefa_big5_results_period %>% 
  specify(승률 ~ 시기) %>%
  hypothesize(null = 'independence') %>%
  generate(reps = 1000, type = 'permute') %>%
  calculate(stat = 't',
            order = c('AC', 'BC')) %>% 
  visualize(method = 'both') +
  shade_p_value(obs_stat = -.466, direction = 'less')

# %%
uefa_big5_results_period %>%
  # Pivot to wide format to create a 'difference' column
  tidyr::pivot_wider(names_from = 시기, values_from = 승률) %>%
  mutate(diff = AC - BC) %>%
  # Now run a one-sample t-test on the difference
  t_test(response = diff, mu = 0, alternative = "less")

# %%
uefa_big5_results_period %>% 
  pivot_wider(names_from='시기', values_from='승률') %>% 
  mutate(차이 = AC - BC) %>% 
  summarise(차이_평균 = mean(차이),
            차이_표준편차 = sd(차이),
            차이_표준오차 = 차이_표준편차 / sqrt(78))     

# %%
-.0176 / .0286

# %%
uefa_big5_results_period %>% 
  specify(승률 ~ 시기) %>%
  hypothesize(null = 'point', mu = 0) %>%
  calculate(stat = 't',
            order = c('AC', 'BC')) %>% 
  visualize(method = 'theoretical') +
  shade_p_value(obs_stat = -.615, direction = 'less')

# %%
'19_20_nba.csv' %>% 
  read.csv() %>% 
  as_tibble -> nba_match_results

# %%
nba_match_results %>% 
  names()

# %%
nba_match_results %>% 
  mutate(
    장소 = 장소 %>% fct_relevel('안방', '방문'),
    시기 = 시기 %>% fct_relevel('BC', 'AC'),
    리그 = 리그 %>% fct_relevel('정규리그', '플레이오프')
  ) -> nba_match_results

# %%
nba_match_results %>%
  filter(리그 != '플레이오프') %>% 
  group_by(팀, 장소, 시기) %>%
  summarise(승률  = mean(승리), .groups = 'drop') %>%
  ggplot(aes(x = 장소, y = 승률)) +
  geom_boxplot() +
  facet_grid(. ~ 시기)

# %%
nba_match_results %>%
  filter(리그 != '플레이오프', 장소 == '안방') %>% 
  group_by(팀, 시기) %>%
  summarise(승률 = mean(승리), .groups = 'drop') %>%
  pivot_wider(names_from = 시기, values_from = 승률) %>%
  # Filter out teams that don't have data for both periods
  drop_na(BC, AC) %>% 
  mutate(diff = BC - AC) %>%
  t_test(response = diff, mu = 0)

# %%
nba_match_results %>%
  filter(리그 != '플레이오프') %>% 
  group_by(팀, 장소, 시기) %>%
  filter(장소 == '안방') %>% 
  summarise(승률  = mean(승리), .groups = 'drop') %>% 
  pivot_wider(names_from = '시기', values_from = '승률') %>% 
  drop_na()

# %%
nba_match_results %>%
  filter(리그 != '플레이오프', 장소 == '안방') %>% 
  group_by(팀, 시기) %>%
  summarise(승률 = mean(승리), .groups = 'drop') %>% 
  pivot_wider(names_from = 시기, values_from = 승률) %>% 
  drop_na(BC, AC) %>% 
  # 1. Create the difference column
  mutate(diff = BC - AC) %>% 
  # 2. Test if the mean difference is significantly different from 0
  t_test(response = diff, mu = 0)

# %%
nba_match_results %>%
  filter(시기 == 'BC') %>% 
  group_by(팀, 장소) %>%
  summarise(승률  = mean(승리), .groups = 'drop') %>% 
  specify(승률 ~ 장소) %>% 
  hypothesize(null = 'independence') %>%
  generate(reps = 1000, type = 'permute') %>%
  calculate(stat = 'diff in means',
            order = c('안방', '방문')) %>%
  select(stat) %>% 
  mutate(type = 'h0') -> nba_simulation_h0

# %%
nba_simulation_h0 %>%
  summarise(low = quantile(stat, .95))

# %%
nba_match_results %>%
  filter(시기 == 'BC') %>% 
  group_by(팀, 장소) %>%
  summarise(승률 = mean(승리),
            .groups = 'drop')  %>% 
  pivot_wider(names_from = '장소',
              values_from = '승률') %>% 
  mutate(차이 = 안방 - 방문) %>% 
  rep_sample_n(reps = 1000, size = 30, replace = TRUE) %>% 
  group_by(replicate) %>% 
  summarise(stat = mean(차이), .groups = 'drop') %>% 
  select(stat) %>% 
  mutate(type = 'h1') -> nba_simulation_h1

# %%
bind_rows(
  nba_simulation_h0,
  nba_simulation_h1
) %>% 
  ggplot(aes(x=stat, fill = type)) +
  geom_histogram(color = 'white',  
                 binwidth = .01, alpha = .5,
                 position = 'identity') +
  geom_vline(xintercept = .0747,
             linetype = 'dashed', lwd=1)

# %%
nba_simulation_h1 %>% 
  filter(stat > .0747) %>% 
  tally()

# %%
nba_match_results %>%
  filter(시기 == 'BC') %>% 
  group_by(팀, 장소) %>% 
  summarise(승률 = mean(승리), .groups = 'drop') %>% 
  # 1. Move '안방' and '방문' into their own columns
  tidyr::pivot_wider(names_from = 장소, values_from = 승률) %>% 
  # 2. Remove teams that don't have both Home and Away data (if any)
  tidyr::drop_na(안방, 방문) %>% 
  # 3. Calculate the difference
  mutate(diff = 안방 - 방문) %>% 
  # 4. Use the one-sample t_test on the difference
  t_test(response = diff, mu = 0)

# %%
pacman::p_load(pwr)

args(power.t.test)

# %%
nba_match_results %>%
  filter(시기 == 'BC') %>%
  group_by(팀, 장소) %>%
  summarise(승률 = mean(승리), .groups = 'drop') %>%
  group_by(장소) %>%
  summarise(승률 = mean(승률), .groups = 'drop')

# %%
nba_match_results %>%
  filter(시기 == 'BC') %>%
  group_by(팀, 장소) %>%
  summarise(승률 = mean(승리), .groups = 'drop') %>%
  pivot_wider(names_from = '장소', values_from = '승률') %>%
  mutate(차이 = 안방 - 방문) %>%
  summarise(sd = sd(차이))

# %%
power.t.test(delta = .106, sd = .158,
             sig.level = .05,
             power = .8,
             type = 'paired',
             alternative = 'one.sided')

# %%
